enhancements/update/distribute-secondary-metadata-as-container-image.md

```diff
@@ -0,0 +1,197 @@
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+
+### Graduation Criteria
+**Note:** *Section not required until targeted at a release.*
+
+- Maturity levels - `Developer Preview`, `Red Hat Staging`, `Red Hat Production`, `GA`
+
+##### `Developer Preview`
+- New Cincinnati plugin
+- Container image registry managed by and dedicated to prow CI
+- Integration tests using an image (to-be-determined: verifying against mocked signatures)
+- CPaaS pipeline is setup up to automatically build and publish a container image for changes to the master branch of the graph-data repository
+
+##### `Developer Preview` -> `Red Hat Staging`
+- Update the Cincinnati deployment template to use the new plugin.
+
+##### Red Hat Staging -> Red Hat Production
+- New plugin is successfully deployed to production
+- Adjust the Cincinnati Operator to use the new plugin
+
+##### Red Hat Production -> GA
+- Document the mirroring process for end-users
+- Definitive positive feedback from the ACM team and their use-case
+
+##### Removing a deprecated feature
+*Not applicable*
+
+### Upgrade / Downgrade Strategy
+*Not applicable*
+
+### Version Skew Strategy
+*Not applicable*
+
+## Implementation History
+
+### Initial implementation
+The initial implementation involves setting up the CPaaS pipeline, implementing a new Cincinnati plugin which downloads the published image and extracts it to disk for subsequent processing in the plugin chain.
+It also involves adjusting the Cincinnati Operator to use the new plugin by default.
+
+## Drawbacks
+
+### Increase OCP release latency
+The enhancement introduces the additional latency of the CPaaS pipeline into the OCP release workflow.
+Assuming that the pipeline is fully automated, the latency introduced is expected to be at the order of minutes.
+This seems to be an acceptable latency with regards to its worst-case scenario, where it's added on top of the duration of blocking an update path via the path of the repository.
+
+## Alternatives
+I want to clearly state that none of the alternatives mentioned below have all the same benefits as the enhancement described here.
+
+Instead of proceeding with this enhancement, we could continue to work with, or rather around, the current implementation of secondary metadata distribution in Cincinnati.
+
+Without any code changes, this would mean that we document the various ways how the cincinnati-graph-data GitHub repository can be mirrored and provided to Cincinnat, and how to reconfigure the plugin settings to directly read the data from disk.
```

Comment:
I think it's worth including more detail on the alternative that is currently implemented with the cincinnati-operator. I wouldn't say it's a better option that this proposal, but since it works and is even implemented, it's worth describing.

It only applies if cincinnati is being run in k8s, which could be a downside if there's desire to run it outside of k8s.

The container image that has the graph data can be used as an init container to cincinnati. So long as it has the `cp` binary or similar, it can copy the data into an `emptyDir` volume, from which cincinnati can read the data.

An advantage is that trust and integrity are handled by the cluster's normal process of running a container.

It's up to the cincinnati-operator in that scenario to restart the pod when a new graph data image should be used. Today, that's determined based [on a field in the `Cincinnati` CRD Spec](https://github.com/openshift/cincinnati-operator/blob/58a04d253067be5f66d584f7f58e8e94898e071e/pkg/apis/cincinnati/v1beta1/cincinnati_types.go#L25-L27).

Reply:
Thanks, I added some of this information to the enhancement.
Is it correct that the responsibility of building the init container image from the graph data on GitHub is with the customer in this case?

Reply:
Yes, the current implementation requires the customer to build the image them self whenever they want a new "snapshot" of the data. It would be an improvement if they could instead just retrieve that image from Red Hat.

Reply:
This enhancement doesn't cover the init container image, but it does cover the graph data image, in fact that is the primary goal.

Reply:
Right. The "Plugin implementation" section discusses one way that the container image can be accessed and consumed. I'm suggesting that in the Alternatives section there be at least some discussion of the init container pattern as a different way to access and consume the image.

Reply:
I added https://github.com/openshift/enhancements/pull/310/files#diff-28b30f8a4731d3d193a2d4f521f39d71R191-R192 in response to your original comment here. PTAL and provide feedback if your comment is satisfied.